app/services/district_report_service.rb

```diff
@@ -20,25 +23,52 @@ def call
     data
   end
 
-  def registrations_count(time)
-    Patient.where(registration_facility: @facilities).with_hypertension.where("recorded_at <= ?", time).count
-  end
-
   def compile_control_and_registration_data
-    @data[:cumulative_registrations] = registrations_count(selected_date.end_of_month)
-    -11.upto(0).each do |n|
+    months_of_data = [registration_counts.to_a.size, MAX_MONTHS_OF_DATA].min
+    @data[:cumulative_registrations] = lookup_registration_count(selected_date)
+    (-months_of_data + 1).upto(0).each do |n|
       time = selected_date.advance(months: n).end_of_month
       formatted_period = time.to_s(:month_year)
 
       @data[:controlled_patients][formatted_period] = controlled_patients(time).count
-      @data[:registrations][formatted_period] = registrations_count(time)
+      @data[:registrations][formatted_period] = lookup_registration_count(time)
     end
   end
 
   def format_quarter(quarter)
     "Q#{quarter.number}-#{quarter.year}"
   end
 
+  def lookup_registration_count(date)
+    lookup_date = date.beginning_of_month.to_date.to_s
+    row = registration_counts.find { |r| r["date"] == lookup_date }
+    return 0 unless row
+    row["running_ct"].to_i
+  end
+
+  def registration_counts
```

Comment:
Could this whole thing just be written as?

```
Facility.first.registered_patients.with_hypertension.group_by_period(:month, :recorded_at, last: 24).count
```

Reply:
One caveat to be aware whilst using `group_by_period` and the likes is that it doesn't generate the in-betweener dates iff the min date in the database is much later than the point in time you're asking it to go back to, for eg:

if you want last 30 days of data, and you only have last 15 days in the database, it won't populate the first 15 days with zeroes.

The way I've typically handled this in the past is that I leave the wholesale date knowledge up to the view; the view knows what it needs to populate; the query knows what it currently has (it doesn't try to backfill) and then you smash the 2 together.

Reply:
> Could this whole thing just be written as?
> 
> ```
> 
> ```

It could, but we would need to add the appropriate association(s) on FacilityGroup to retrieve the counts for the set of facilities inside a facility group. We need these counts at the District (i.e. FacilityGroup) level, not Facility.

I'll add that to a list of followup improvements I'm keeping here: https://www.pivotaltracker.com/story/show/173575455

Reply:
Cool, that sounds good.

Fwiw, if you just need a wholesale count for a `FacilityGroup`, one could do:

```
FacilityGroup.first.patients.with_hypertension.group_by_period(:month, :recorded_at, last: 24).coun
```

Without any code-changes.

It'll be slightly more work if we need to preserve Facility IDs.

Reply:
I realized now why none of these simpler solutions would work. We need a running total from the beginning of time of registered patients in a FacilityGroup. Groupdate makes it easy to get the totals for every month for a particular period, but it doesn't handle the cumulative total bit. 

I'm going to look into getting the counts via groupdate and then totaling via ruby to see how that looks.